services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/Exclude.java

```diff
@@ -0,0 +1,210 @@
+   *
+   * @return a list of strings matching one
+   *   of the {@link DirectionsCriteria.ExcludeCriteria} exclusions
+   */
+  @Nullable
+  public abstract List<String> criteria();
+
+  /**
+   * Exclude certain points from routing. The default is to not exclude anything.
+   *
+   * @return a list of points excluded from the routing.
+   */
+  @Nullable
+  public abstract List<Point> points();
+
+  /**
+   * Use this builder to build an {@link Exclude} object.
+   */
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    /**
+     * Exclude certain road types from routing. The default is to not exclude anything from the
+     * profile selected. The following exclude flags are available for each profile:
+     * <p>
+     * {@link DirectionsCriteria#PROFILE_DRIVING}: One of {@link DirectionsCriteria#EXCLUDE_TOLL},
+     * {@link DirectionsCriteria#EXCLUDE_MOTORWAY}, or {@link DirectionsCriteria#EXCLUDE_FERRY}.
+     * <p>
+     * {@link DirectionsCriteria#PROFILE_DRIVING_TRAFFIC}: One of
+     * {@link DirectionsCriteria#EXCLUDE_TOLL}, {@link DirectionsCriteria#EXCLUDE_MOTORWAY}, or
+     * {@link DirectionsCriteria#EXCLUDE_FERRY}.
+     * <p>
+     * {@link DirectionsCriteria#PROFILE_WALKING}: No excludes supported
+     * <p>
+     * {@link DirectionsCriteria#PROFILE_CYCLING}: {@link DirectionsCriteria#EXCLUDE_FERRY}
+     *
+     * @param criteria a list of exclude criteria
+     *   matching {@link DirectionsCriteria.ExcludeCriteria}
+     * @return this builder.
+     */
+    public abstract Builder criteria(@Nullable List<String> criteria);
+
+    /**
+     * Exclude certain points from routing. The default is to not exclude anything.
+     *
+     * @param points a list of points to exclude.
+     * @return this builder.
+     */
+    public abstract Builder points(@Nullable List<Point> points);
+
+    /**
+     * Builds the object.
+     *
+     * @return a new instance of {@link Exclude}
+     */
+    public abstract Exclude build();
+  }
+
+  @Nullable
+  static Exclude fromUrlQueryParameter(@Nullable String queryParameter) {
+    if (queryParameter == null || Objects.equals(queryParameter, "")) {
+      return null;
+    }
+
+    ArrayList<Point> points = new ArrayList<>();
+    ArrayList<String> criteria = new ArrayList<>();
+
+    List<String> excludeItems = ParseUtils.parseToStrings(queryParameter, ",");
+    if (excludeItems != null) {
+      for (String excludeItem : excludeItems) {
+        if (excludeItem.startsWith("point(") && excludeItem.endsWith(")")) {
+          Point point = parsePoint(excludeItem);
+          points.add(point);
+        } else if (excludeItem.contains("(") && excludeItem.contains(")")) {
+          // Ignoring the unexpected type of data. Update the library to get support of new types
+        } else if (excludeItem.contains("(") || excludeItem.contains(")")) {
+          throw new IllegalArgumentException("Can't parse parameter " + excludeItem);
```

Comment:
Let's remove these 2 cases and let the backend respond with an appropriate errors. We should only handle what we expect.

Reply:
Sorry, now I see it's the other way around. Regardless, since the server-side API format is generic string, we cannot throw an exception on the parsing level ourselves. We probably need to remove that if-check.

Reply:
this line is about a case when you put a comma instead of the space in point
```kotlin
  @Test(expected = IllegalArgumentException.class)
  public void parsePointWithCommaDelimiter() {
    String invalidPoint = "point(5.0,7.0)";
    Exclude.fromUrlQueryParameter(invalidPoint);
  }
``` 
Do you suggest to ignore everything we can't parse? Like:
```kotlin
  @Test
  public void parsePointWithCommaDelimiter() {
    String invalidPoint = "point(5.0,7.0)";
    Exclude exclude = Exclude.fromUrlQueryParameter(invalidPoint);
    assertNull(exclude);
  }
```

Reply:
Yeah, I agree with @LukasPaczos providing the right format should be managed by the end-user when integrating the API and check on the server side. I believe we should also remove the previous `if` https://github.com/mapbox/mapbox-java/pull/1362#discussion_r808503881 If in the future Directions API adds new data formats for exclude, we'll accommodate [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

Reply:
@LukasPaczos , @Guardiola31337 , updated parsing with a new concept: `Exclude` is just a type-safe way of working with known excludes.  All the unknown properties and flags are ignored now

Reply:
Why even ignore it and return `null` from our side? Why can't we pass whatever we get from the client (assuming is well formed / formatted and let the server crash if not? https://github.com/mapbox/mapbox-java/blob/f2c30793a4054248bdb95bf511d53a3da2256662/services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/Exclude.java#L184

Reply:
@Guardiola31337 , the line 184 is for reading points, not for writing. You can write incorrect point like `exclude("point(test)")`, but you won't be able to read it as a `Point` because it can't be parsed